src/test_util.rs

```diff
@@ -161,14 +162,53 @@ pub fn lines_from_file<P: AsRef<Path>>(path: P) -> Vec<String> {
 }
 
 pub fn wait_for(f: impl Fn() -> bool) {
-    let wait = std::time::Duration::from_millis(5);
-    let limit = std::time::Duration::from_secs(5);
-    let mut attempts = 0;
-    while !f() {
-        std::thread::sleep(wait);
-        attempts += 1;
-        if attempts * wait > limit {
-            panic!("timed out while waiting");
+    WaitFor::is(f).now()
+}
+
+pub struct WaitFor<T, F> {
+    f: F,
+    limit: Duration,
+    wait: Duration,
+    _marker: std::marker::PhantomData<T>,
+}
+
+impl WaitFor<(), ()> {
+    pub fn is(mut predicate: impl FnMut() -> bool) -> WaitFor<(), impl FnMut() -> Option<()>> {
+        WaitFor::it(move || if predicate() { Some(()) } else { None })
+    }
+}
+
+impl<T, F: FnMut() -> Option<T>> WaitFor<T, F> {
+    pub fn it(f: F) -> Self {
```

Comment:
There is probably a better name for this, `it` doesn't seem very clear to me.

Reply:
`data` could be clearer, also `some`. 

`WaitFor::some(...)` should be the clearest.